li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java

```diff
@@ -0,0 +1,348 @@
+  private UUID _clientId;
+
+  // Per cluster Consumers
+  private Map<ClusterDescriptor, LiKafkaConsumer<K, V>> _consumers;
+
+  // Consumer builder for creating per-cluster LiKafkaConsumer
+  private LiKafkaConsumerBuilder<K, V> _consumerBuilder;
+
+  // Consumer configs common to all clusters
+  private LiKafkaConsumerConfig _commonConsumerConfigs;
+
+  public LiKafkaFederatedConsumerImpl(Properties props) {
+    this(new LiKafkaConsumerConfig(props), null, null);
+  }
+
+  public LiKafkaFederatedConsumerImpl(Properties props, MetadataServiceClient mdsClient, LiKafkaConsumerBuilder<K, V> consumerBuilder) {
+    this(new LiKafkaConsumerConfig(props), mdsClient, consumerBuilder);
+  }
+
+  public LiKafkaFederatedConsumerImpl(Map<String, Object> configs) {
+    this(new LiKafkaConsumerConfig(configs), null, null);
+  }
+
+  public LiKafkaFederatedConsumerImpl(Map<String, Object> configs, MetadataServiceClient mdsClient,
+      LiKafkaConsumerBuilder<K, V> consumerBuilder) {
+    this(new LiKafkaConsumerConfig(configs), mdsClient, consumerBuilder);
+  }
+
+  @SuppressWarnings("unchecked")
+  private LiKafkaFederatedConsumerImpl(LiKafkaConsumerConfig configs, MetadataServiceClient mdsClient,
+      LiKafkaConsumerBuilder<K, V> consumerBuilder) {
+    _commonConsumerConfigs = configs;
+    _clusterGroup = new ClusterGroupDescriptor(configs.getString(LiKafkaConsumerConfig.CLUSTER_ENVIRONMENT_CONFIG),
+        configs.getString(LiKafkaConsumerConfig.CLUSTER_GROUP_CONFIG));
+
+    // Each per-cluster consumer and auditor will be instantiated by the passed-in consumer builder when the client
+    // begins to consume from that cluster. If a null builder is passed, create a default one, which builds
+    // LiKafkaConsumer.
+    _consumers = new ConcurrentHashMap<ClusterDescriptor, LiKafkaConsumer<K, V>>();
+    _consumerBuilder = consumerBuilder != null ? consumerBuilder : new LiKafkaConsumerBuilder<K, V>();
+
+    try {
+      // Instantiate metadata service client if necessary.
+      _mdsClient = mdsClient != null ? mdsClient :
+          configs.getConfiguredInstance(LiKafkaConsumerConfig.METADATA_SERVICE_CLIENT_CLASS_CONFIG,
+              MetadataServiceClient.class);
+
+      // Register this federated client with the metadata service. The metadata service will assign a UUID to this
+      // client, which will be used for later interaction between the metadata service and the client.
+      //
+      // Registration may also return further information such as the metadata server version and any protocol settings.
+      // We assume that such information will be kept and used by the metadata service client itself.
+      _clientId = mdsClient.registerFederatedClient(_clusterGroup, configs.originals());
+    } catch (Exception e) {
+      try {
+        if (_mdsClient != null) {
+          _mdsClient.close();
+        }
+      } catch (Exception e2) {
+        LOG.error("Failed to close the metadata service client", e2);
```

Comment:
easier to just do 2.addSuppresed(e2) - would also cause the original exception to be printed 1st (more usable)

Reply:
Done